### DISCUSSION TITLE
Empty for loop for non-empty array

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#4348] [1. Help] [answered]
"Top Mitwirkende" is always empty, but the array `persons` does actually have items in it. How can this be? I'm really going crazy here...

![image](https://github.com/user-attachments/assets/e579ee22-64d6-4f58-ae7c-e59bc7d88d0a)

![image](https://github.com/user-attachments/assets/ac6479ae-d9de-4882-8711-77e5b2f78a0c)

**Comment by ekwoka** [comment]
What does the Dom look like?

**Reply by simonferndriger** [reply]
The template does not render in the dom:

![image](https://github.com/user-attachments/assets/ef21ed2c-bcfc-42d7-8263-9a569d8f9b4a)

What is also interesting: this is an ajax-include. When I use this include on ther html pages, it works!

**Comment by SimoTod** [comment]
Not sure if it's related but you have a function (well, an alpine data generator l) and a variable with the
same name (persons) which is probably confusing for Alpine and the future
yourself looking back at the code.

**Reply by simonferndriger** [reply]
Yes you are right, I actually asked myself the same thing - is that an issue - or not?

**Reply by simonferndriger** [reply]
PS: I explicetly renamed the array to `people` but it still didn't work...

**Reply by ekwoka** [reply]
Never seen anything like this happen....definitely can't diagnose from here.

What does does the update?

the name conflict wouldn't an issue unless it was defined above the x-data using the name.

But it certainly makes it confusing.

**Reply by simonferndriger** [reply]
If you want to, you can see it online.

On this page, it works just fine:
https://matchflix.ch/?movie=704239

Here it doesn't:
https://matchflix.ch/discover/ > Search "The Union" > Inspect

**Reply by simonferndriger** [reply]
Or to be a bit more precise:
https://matchflix.ch/discover/ > Search "The Union" > Click on the movie poster > Click in the upper right corner on the inspect/search icon

This will search for movie details (credits, similar movies) while the credits (i.e. "Top Mitwirkende") won't show up.

**Reply by SimoTod** [reply]
<img width="1282" alt="Screenshot 2024-08-23 at 00 34 50" src="https://github.com/user-attachments/assets/fad6d428-0689-4b83-88e4-e6e8419946b3">

If I inspect the page you linked after those steps, the persons array is empty so it works as expected. I think the code is supposed to populate the list in a $watch callback so, if you are expecting results, perhaps you are setting the variable to watch before you create the watcher.

**Reply by simonferndriger** [reply]
Thanks! I didn't know the inspector can do that - in that case I will have to look for a reason the array get's reset to being empty I guess.

**Reply by simonferndriger** [reply]
@SimoTod But even if I would set the watcher AFTER I set the variable, if it shows that persons is populated, it should still render in the template! No matter if it was empty before. So, to be able to catch as many changes to persons as possible, I located the watcher in the init() function of the persons() component.

I even documented out ALL possible ways that persons could be reset and still the same result (which is easy, since it can only be alterted from WITHIN the component and not from outside, it's a small component). So, yes, the template loop seems to work as expected, BUT I really think there is a bug that persons is empty even thoug it shouldn't be.  See this console output as an example (which is live by now as well at https://matchflix.ch/discover/):

![image](https://github.com/user-attachments/assets/8c0de15a-b8cb-46bc-ad04-d2cee021d02f)

![image](https://github.com/user-attachments/assets/b11b0013-8e17-43eb-81bb-0a8bf0d277de)

This can't be explained, right?

**Reply by SimoTod** [reply]
<img width="764" alt="Screenshot 2024-08-23 at 12 44 24" src="https://github.com/user-attachments/assets/7bd41fd2-64c7-4f59-ac9b-6d50a9058727">
There is something in your code that sets the list back to []. See screenshot

**Reply by simonferndriger** [reply]
OK, I'm really confused now. If I look at it in incognito mode, I see the same as you (3x empty). But when I'm looking at it with all my localstorage data, I see a full array but an empty template... how can this be?

**Reply by ekwoka** [reply, marked as answer]
Sounds like you might be having something set the data to the underlying object and not the proxy.

So when Alpine accesses it doesn't have data, but you add it to the target afterwards, but Alpine can't react to that. You see it filled in the console because the console is lazily evaluated

**Reply by simonferndriger** [reply]
Ah, OK - yes, the confusing thing was that I saw it sometimes in the console, but other times not. Glad you could put that in perspective! 🙏

**Comment by simonferndriger** [comment]
I use the same component twice on the same page (once for the cast, once for the person search), do they share the SAME attribute values or do they have individual ones for each component instance? Because the persons for the search is indeed empty. It could then be a timing problem (search is finishing after the inspect).

**Reply by simonferndriger** [reply]
Ah, I got it - the ajax was loaded conditionally "x-ajax="search.active && '../ajax/search.html?' + version", so when search.active was false for a short amount of time, this was resetting it...

**Comment by simonferndriger** [comment]
Thank you so much for looking into this 🙏
